### ISSUE TITLE
NPM Dependency NestJS for cloud.HttpServer causes Stack Overflow

### ISSUE THREAD

**Issue by WonderPanda** [pulumi/pulumi#2031]
First of all thanks so much for your work on Pulumi. It's been blowing me away so far!

I was hoping to be able to run a NestJS application using the cloud.HttpServer abstraction since it essentially boils down to a fancy wrapper around an Express instance. In my proof of concept as soon as I add the `@nestjs/common` package to my Pulumi app the Stack can no longer be deployed and errors out with the following:
```
Diagnostics:
  pulumi:pulumi:Stack: aws-typescript-test-test
    error: Running program '/home/jesse/code/pulumi-tests/aws-typescript-test' failed with an unhandled exception:

    error: RangeError: Maximum call stack size exceeded
        at findDependency (/home/jesse/code/pulumi-tests/aws-typescript-test/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:223:46)
        at addPackageAndDependenciesToSet (/home/jesse/code/pulumi-tests/aws-typescript-test/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:176:19)
        at recurse (/home/jesse/code/pulumi-tests/aws-typescript-test/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:205:17)
        at addPackageAndDependenciesToSet (/home/jesse/code/pulumi-tests/aws-typescript-test/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:199:9)
        at recurse (/home/jesse/code/pulumi-tests/aws-typescript-test/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:205:17)
        at addPackageAndDependenciesToSet (/home/jesse/code/pulumi-tests/aws-typescript-test/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:199:9)
        at recurse (/home/jesse/code/pulumi-tests/aws-typescript-test/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:205:17)
        at addPackageAndDependenciesToSet (/home/jesse/code/pulumi-tests/aws-typescript-test/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:199:9)
        at recurse (/home/jesse/code/pulumi-tests/aws-typescript-test/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:205:17)
        at addPackageAndDependenciesToSet (/home/jesse/code/pulumi-tests/aws-typescript-test/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:199:9)

    error: an unhandled error occurred: Program exited with non-zero exit code: 1

error: an error occurred while advancing the preview
```
I've been looking at the source for pulumi/runtime/closure/codepaths but am still not familiar enough to really tell what would cause this issue. If you have any thoughts please let me know, I think the ability to run more complex Node applications using cloud.HttpServer would be a killer feature.

**Comment by CyrusNajmabadi** [comment]
Thanks for the report!  will look asap.  Just out of curiosity, is there a specific version of "@nestjs/common" you reference in your package.json?  It may or may not be relevant.  And i'd like to make sure i'm doing exacty what you did.

> If you have any thoughts please let me know, I think the ability to run more complex Node applications using cloud.HttpServer would be a killer feature.

Bsaically, what Pulumi does is walk your package.json dependency tree to figure out all the right packages to upload to aws/azure.  It seems like there's a case here were because of how things are authored, there's a circularity in the package.json dependency structure, and that's causing us to blow the stack.  I really need to take a look at that package to understand what's going and what the right fix might be.  A simple fix would probably be to just have some mechanism that prevents recursion.  But i'm surprised recursion can even happen in the first place in npm because i have no idea how those packages would work in the first place! :)

**Comment by WonderPanda** [comment]
Hey @CyrusNajmabadi sorry for the delay the start of this week has been super busy. The version of @nestjs/common I'm resolving to right now is 5.3.11. I create a repo that reproduces the issue:
https://github.com/WonderPanda/pulumi-nestjs

The idea here ultimately is to be able to use the nestjs wrapper around express to power the cloud.HttpServer but that also brings up a separate issue around async initialization of the app instance. From what I've encountered so far it seems like the initialization needs to happen inside of the `cloud.RequestListenerFactory` (which makes sense) so it's not clear if what I'm trying is even possible. Are there plans to bring a Promise based implementation to the factory?

**Comment by CyrusNajmabadi** [comment]
> Hey @CyrusNajmabadi sorry for the delay the start of this week has been super busy. The version of @nestjs/common I'm resolving to right now is 5.3.11. I created a repo that reproduces the issue:
https://github.com/WonderPanda/pulumi-nestjs. Running pulumi up in the root of the project generates the stack overflow issue:

Great, i'll check that out in the next few days!

**Comment by CyrusNajmabadi** [comment]
> Are there plans to bring a Promise based implementation to the factory?

Interesting... i'm not exactly certain how that would actually work.  We need to be able to provide a real javascript callback function as the export of the AWS Lambda.  It's unclear how we would do that asynchronously.  Do you have a sort of codgen in mind that would work?

**Comment by CyrusNajmabadi** [comment]
Hey @WonderPanda can upload your package.json file?  It would be very helpful.  Thanks!

**Comment by CyrusNajmabadi** [comment]
weird.  es6-iterator depends on es5-ext, which depends on es6-iterator

I guess circular packages just exist in the wild.  Will add a circularity checker to our path computation algorithm.  Thanks for reporting this!
